src/net/server/mod.rs

```diff
@@ -216,6 +246,249 @@ where
+            // No client identity will be associated with the connection if:
+            //  * No certificate was supplied.
+            //  * There was a problem interpreting the certificate. It is unlikely to see an invalid
+            //    certificate here, because the certificate must have passed full verification at
+            //    connection time. But it's possible the certificate subject is not something we
+            //    recognize as a helper.
+            let id = Self::identify_client(
+                &network_config,
+                stream
+                    .get_ref()
+                    .1
+                    .peer_certificates()
+                    .and_then(<[_]>::first),
+            );
+            let service = SetClientIdentityFromCertificate { inner: service, id };
+            Ok((stream, service))
+        })
+    }
+}
+
+#[derive(Clone)]
+struct SetClientIdentityFromCertificate<S> {
+    inner: S,
+    id: Option<ClientIdentity>,
+}
+
+impl<B, S: Service<Request<B>>> Service<Request<B>> for SetClientIdentityFromCertificate<S> {
+    type Response = S::Response;
+    type Error = S::Error;
+    type Future = S::Future;
+
+    #[inline]
+    fn poll_ready(&mut self, cx: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
+        self.inner.poll_ready(cx)
+    }
+
+    fn call(&mut self, mut req: Request<B>) -> Self::Future {
+        if let Some(id) = self.id {
+            req.extensions_mut().insert(id);
+        }
+        self.inner.call(req)
+    }
+}
+
+/// Name of the header that passes the client identity when not using HTTPS.
+pub static HTTP_CLIENT_ID_HEADER: HeaderName =
+    HeaderName::from_static("x-unverified-client-identity");
+
+/// Service wrapper that gets a client helper identity from a header.
+///
+/// Since this allows a client to claim any identity, it is completely
+/// insecure. It must only be used in contexts where that is acceptable.
+#[derive(Clone)]
+pub(super) struct SetClientIdentityFromHeader<S> {
+    inner: S,
+}
+
+impl<S> SetClientIdentityFromHeader<S> {
+    fn new(inner: S) -> Self {
+        Self { inner }
+    }
+}
+
+impl<B, S: Service<Request<B>, Response = Response>> Service<Request<B>>
+    for SetClientIdentityFromHeader<S>
+{
+    type Response = Response;
+    type Error = S::Error;
+    type Future = Either<S::Future, Ready<Result<Response, S::Error>>>;
+
+    #[inline]
+    fn poll_ready(&mut self, cx: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
+        self.inner.poll_ready(cx)
+    }
+
+    fn call(&mut self, mut req: Request<B>) -> Self::Future {
+        if let Some(header_value) = req.headers().get(HTTP_CLIENT_ID_HEADER.clone()) {
+            let id_result = header_value
+                .to_str()
+                .map_err(Into::into)
+                .and_then(|value_str| usize::from_str(value_str).map_err(Into::into))
+                .and_then(|value_int| {
+                    HelperIdentity::try_from(value_int).map_err(|e| {
```

Comment:
maybe we can add `HelperIdentity::new(&str)` constructor later. Id does not really have to be an integer

Reply:
I changed it to use serde consistently for `HelperIdentity`. It seems a little weird to use `serde_json` to do integer-string conversions, but the serde support for this type is necessary anyways since it appears in a `PrepareQuery` message.